frostsnap_core/src/device.rs

```diff
@@ -403,105 +444,97 @@ impl FrostSigner {
-                    DeviceSend::ToStorage(DeviceToStorageMessage::ExpendNonce),
+                    DeviceSend::ToStorage(DeviceToStorageMessage::ExpendNonce {
+                        nonce_counter: self.nonce_counter,
+                    }),
                     DeviceSend::ToCoordinator(DeviceToCoordinatorMessage::SignatureShare {
                         signature_shares,
-                        new_nonces: replenish_nonces,
+                        new_nonces,
                     }),
                 ])
             }
             _ => Err(ActionError::WrongState {
-                in_state: self.state.name(),
+                in_state: self.action_state_name(),
                 action: "sign_ack",
             }),
         }
     }
 
-    pub fn frost_key(&self) -> Option<&FrostKey<Normal>> {
-        match self.state() {
-            SignerState::Registered => None,
-            SignerState::KeyGen { .. } => None,
-            SignerState::FrostKey { key, .. } => Some(&key.frost_key),
-            SignerState::AwaitingSignAck { key, .. } => Some(&key.frost_key),
-        }
+    pub fn action_state_name(&self) -> &'static str {
+        self.action_state
+            .as_ref()
+            .map(|x| x.name())
+            .unwrap_or("None")
     }
 }
 
 #[derive(Clone, Debug, bincode::Encode, bincode::Decode)]
 pub enum SignerState {
-    Registered,
     KeyGen {
         scalar_poly: Vec<Scalar>,
         device_to_share_index: BTreeMap<DeviceId, Scalar<Public, NonZero>>,
-        threshold: usize,
-        aux_rand: [u8; 32],
+        threshold: u16,
+    },
+    KeyGenAck {
+        key: FrostsnapSecretKey,
     },
     AwaitingSignAck {
-        key: FrostsnapKey,
+        key_id: KeyId,
         sign_task: SignTask,
-        nonces: BTreeMap<Scalar<Public, NonZero>, (Vec<Nonce>, usize, usize)>,
-    },
-    FrostKey {
-        key: FrostsnapKey,
-        awaiting_ack: bool,
+        session_nonces: BTreeMap<Scalar<Public, NonZero>, SignRequestNonces>,
     },
 }
 
 impl SignerState {
     pub fn name(&self) -> &'static str {
         match self {
-            SignerState::Registered => "Registered",
             SignerState::KeyGen { .. } => "KeyGen",
+            SignerState::KeyGenAck { .. } => "KeyGenAck",
             SignerState::AwaitingSignAck { .. } => "AwaitingSignAck",
-            SignerState::FrostKey { .. } => "FrostKey",
         }
     }
 }
 
 #[derive(Clone, Debug, bincode::Encode, bincode::Decode)]
-pub struct FrostsnapKey {
+pub struct FrostsnapSecretKey {
     /// The joint key
     pub frost_key: FrostKey<Normal>,
     /// Our secret share of it
     pub secret_share: Scalar,
     /// Our secret share index
     pub share_index: Scalar<Public, NonZero>,
-    /// auxilliary randomness for generating nonces
-    pub aux_rand: [u8; 32],
+}
+
+impl FrostsnapSecretKey {
+    pub fn key_id(&self) -> KeyId {
+        self.frost_key.key_id()
+    }
 }
 
 #[derive(Debug, Clone)]
 pub enum StartSignError {
```

Comment:
I think this should be in `coordinator.rs`? Moved in previous PR

Realised because `StartSignError::DeviceNotPartOfKey` doesn't 't make sense from the perspective of the device, previously we discussed that devices do not need to have knowledge of other device ids.

Reply:
good catch!